File: scripts/convert-conversion-time-brasilia-to-chicago.py
#!/usr/bin/env python3
"""
Converte Conversion Time de Brasília (GMT-3) para Chicago (GMT-6)
para upload de conversões offline no Google Ads.

Uso:
    python scripts/convert-conversion-time-brasilia-to-chicago.py arquivo.csv
    python scripts/convert-conversion-time-brasilia-to-chicago.py arquivo.csv -o saida.csv

Colunas esperadas: GCLID, Conversion Name, Conversion Time
Saída: yyyy-MM-dd HH:mm:ss (compatível com Google Sheets/CSV)
"""

import csv
import re
import sys
from datetime import datetime

# Python 3.9+ tem zoneinfo. Para 3.8: pip install backports.zoneinfo
try:
    from zoneinfo import ZoneInfo
except ImportError:
    from backports.zoneinfo import ZoneInfo

BRASILIA = ZoneInfo("America/Sao_Paulo")
CHICAGO = ZoneInfo("America/Chicago")


def parse_brasilia_time(value: str) -> datetime | None:
    """Interpreta a string como horário de Brasília."""
    if not value or not value.strip():
        return None
    value = value.strip()

    # Remove timezone suffix se existir (-03:00, -0300, Z, etc)
    value_clean = re.sub(r"[-+]\d{2}:?\d{2}$", "", value, flags=re.I)
    value_clean = re.sub(r"Z$", "", value_clean, flags=re.I)
    value_clean = value_clean.replace("T", " ").strip()

    formats = [
        "%Y-%m-%d %H:%M:%S",
        "%Y-%m-%d %H:%M",
        "%Y-%m-%d",
        "%d/%m/%Y %H:%M:%S",
        "%d/%m/%Y %H:%M",
        "%d/%m/%Y",
    ]

    for fmt in formats:
        try:
            dt = datetime.strptime(value_clean, fmt)
            return dt.replace(tzinfo=BRASILIA)
        except ValueError:
            continue

    # Fallback: dateutil ou datetime.fromisoformat
    try:
        dt = datetime.fromisoformat(value.replace("Z", "+00:00"))
        if dt.tzinfo is None:
            dt = dt.replace(tzinfo=BRASILIA)
        return dt
    except ValueError:
        pass

    return None


def convert_to_chicago(dt: datetime) -> str:
    """Converte para Chicago e formata como yyyy-MM-dd HH:mm:ss"""
    chicago_dt = dt.astimezone(CHICAGO)
    return chicago_dt.strftime("%Y-%m-%d %H:%M:%S")


def convert_brasilia_to_chicago(value: str) -> str:
    """Converte string em Brasília para string em Chicago."""
    dt = parse_brasilia_time(value)
    if dt is None:
        return ""
    return convert_to_chicago(dt)


def find_column_index(headers: list[str], names: list[str]) -> int:
    """Encontra índice da coluna por nome (case-insensitive)."""
    lower_headers = [h.lower() for h in headers]
    for name in names:
        for i, h in enumerate(lower_headers):
            if name.lower() in h:
                return i
    return -1


def main():
    args = [a for a in sys.argv[1:] if not a.startswith("-")]
    opts = sys.argv[1:]
    input_path = args[0] if args else None
    output_path = None
    if "-o" in opts:
        idx = opts.index("-o")
        if idx + 1 < len(opts):
            output_path = opts[idx + 1]

    if not input_path:
        print("Uso: python convert-conversion-time-brasilia-to-chicago.py <arquivo.csv> [-o saida.csv]")
        sys.exit(1)

    try:
        with open(input_path, "r", encoding="utf-8-sig") as f:
            sample = f.read(4096)
            f.seek(0)
            dialect = csv.Sniffer().sniff(sample, delimiters=",;\t")
            reader = csv.reader(f, dialect)
            rows = list(reader)
    except FileNotFoundError:
        print(f"Arquivo não encontrado: {input_path}")
        sys.exit(1)

    if not rows:
        print("Arquivo vazio.")
        sys.exit(0)

    headers = rows[0]
    conv_time_idx = find_column_index(headers, ["conversion time", "conversion_time"])
    if conv_time_idx < 0:
        print(f'Coluna "Conversion Time" não encontrada. Colunas: {", ".join(headers)}')
        sys.exit(1)

    out_rows = [headers]
    converted = 0
    for row in rows[1:]:
        if len(row) <= conv_time_idx:
            row.extend([""] * (conv_time_idx - len(row) + 1))
        raw = row[conv_time_idx] if conv_time_idx < len(row) else ""
        new_time = convert_brasilia_to_chicago(raw)
        if new_time:
            row[conv_time_idx] = new_time
            converted += 1
        out_rows.append(row)

    if output_path:
        with open(output_path, "w", encoding="utf-8", newline="") as f:
            writer = csv.writer(f, dialect=dialect)
            writer.writerows(out_rows)
        print(f"✅ {converted} conversões processadas. Salvo em: {output_path}")
    else:
        writer = csv.writer(sys.stdout, dialect=dialect)
        writer.writerows(out_rows)


if __name__ == "__main__":
    main()

File: scripts/convert-conversion-time-brasilia-to-chicago.js
#!/usr/bin/env node
/**
 * Converte Conversion Time de Brasília (GMT-3) para Chicago (GMT-6)
 * para upload de conversões offline no Google Ads.
 *
 * Uso:
 *   node scripts/convert-conversion-time-brasilia-to-chicago.js <arquivo.csv>
 *   node scripts/convert-conversion-time-brasilia-to-chicago.js <arquivo.csv> -o saida.csv
 *
 * Colunas esperadas: GCLID, Conversion Name, Conversion Time
 * Saída: yyyy-MM-dd HH:mm:ss (compatível com Google Sheets/CSV)
 */

const fs = require('fs');
const path = require('path');

/**
 * Interpreta a string como horário de Brasília (GMT-3) e retorna o equivalente em Chicago (GMT-6).
 * Brasília está 3h à frente de Chicago → subtrai 3 horas.
 * @param {string} input - Data/hora em qualquer formato comum (ISO, yyyy-MM-dd HH:mm:ss, com/sem timezone)
 * @returns {string} Formato yyyy-MM-dd HH:mm:ss em horário de Chicago
 */
function convertBrasiliaToChicago(input) {
  if (!input || typeof input !== 'string') return '';

  let d;
  const trimmed = input.trim();

  // Se já tem timezone (ex: -03:00, -0300, Z), o Date parseia corretamente
  if (/[-+]\d{2}:?\d{2}$|Z$/i.test(trimmed)) {
    d = new Date(trimmed);
  } else if (/T/.test(trimmed)) {
    // ISO com T
    d = new Date(trimmed);
  } else {
    // Assume Brasília: yyyy-MM-dd HH:mm:ss ou similar
    const brasiliaStr = trimmed.replace(/\s+/, 'T') + '-03:00';
    d = new Date(brasiliaStr);
  }

  if (isNaN(d.getTime())) return '';

  // Formatar em horário de Chicago (America/Chicago)
  const chicago = d.toLocaleString('sv-SE', {
    timeZone: 'America/Chicago',
    hour12: false,
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit'
  });

  // sv-SE retorna "yyyy-MM-dd HH:mm:ss" (ou "yyyy-mm-dd, hh:mm:ss" - verificar)
  return chicago.replace(',', ' ').replace(/\//g, '-');
}

/**
 * Formato alternativo mais explícito para garantir yyyy-MM-dd HH:mm:ss
 */
function formatChicagoTime(date) {
  const parts = new Intl.DateTimeFormat('en-CA', {
    timeZone: 'America/Chicago',
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    second: '2-digit',
    hour12: false
  }).formatToParts(date);

  const get = (type) => parts.find((p) => p.type === type)?.value ?? '00';
  const year = get('year');
  const month = get('month');
  const day = get('day');
  const hour = get('hour');
  const minute = get('minute');
  const second = get('second');
  return `${year}-${month}-${day} ${hour}:${minute}:${second}`;
}

function convertBrasiliaToChicagoStrict(input) {
  if (!input || typeof input !== 'string') return '';
  const trimmed = input.trim();

  let d;
  if (/[-+]\d{2}:?\d{2}$|Z$/i.test(trimmed) || /T/.test(trimmed)) {
    d = new Date(trimmed);
  } else {
    d = new Date(trimmed.replace(/\s+/, 'T') + '-03:00');
  }

  if (isNaN(d.getTime())) return '';
  return formatChicagoTime(d);
}

/**
 * Parse simples de CSV (suporta vírgula e ponto-e-vírgula)
 */
function parseCSV(content) {
  const lines = content.split(/\r?\n/).filter((l) => l.trim());
  if (lines.length === 0) return { headers: [], rows: [] };

  const sep = lines[0].includes(';') ? ';' : ',';
  const headers = lines[0].split(sep).map((h) => h.trim().replace(/^["']|["']$/g, ''));

  const rows = [];
  for (let i = 1; i < lines.length; i++) {
    const values = [];
    let current = '';
    let inQuotes = false;
    for (let j = 0; j < lines[i].length; j++) {
      const c = lines[i][j];
      if (c === '"' || c === "'") inQuotes = !inQuotes;
      else if (!inQuotes && (c === sep || c === '\t')) {
        values.push(current.trim());
        current = '';
      } else {
        current += c;
      }
    }
    values.push(current.trim());
    rows.push(values);
  }

  return { headers, rows };
}

function findColumnIndex(headers, possibleNames) {
  const lower = headers.map((h) => h.toLowerCase());
  for (const name of possibleNames) {
    const idx = lower.findIndex((h) => h.includes(name.toLowerCase()));
    if (idx >= 0) return idx;
  }
  return -1;
}

function main() {
  const args = process.argv.slice(2);
  const inputPath = args.find((a) => !a.startsWith('-'));
  const outIdx = args.indexOf('-o');
  const outputPath = outIdx >= 0 ? args[outIdx + 1] : null;

  if (!inputPath) {
    console.error('Uso: node convert-conversion-time-brasilia-to-chicago.js <arquivo.csv> [-o saida.csv]');
    process.exit(1);
  }

  const fullPath = path.resolve(inputPath);
  if (!fs.existsSync(fullPath)) {
    console.error('Arquivo não encontrado:', fullPath);
    process.exit(1);
  }

  const content = fs.readFileSync(fullPath, 'utf-8');
  const { headers, rows } = parseCSV(content);

  const gclidIdx = findColumnIndex(headers, ['gclid', 'google click id']);
  const convNameIdx = findColumnIndex(headers, ['conversion name', 'conversion_name']);
  const convTimeIdx = findColumnIndex(headers, ['conversion time', 'conversion_time']);

  if (convTimeIdx < 0) {
    console.error('Coluna "Conversion Time" não encontrada. Colunas:', headers.join(', '));
    process.exit(1);
  }

  const sep = content.includes(';') ? ';' : ',';
  const outLines = [headers.join(sep)];

  let converted = 0;
  for (const row of rows) {
    const newRow = [...row];
    const raw = row[convTimeIdx] || '';
    const convertedTime = convertBrasiliaToChicagoStrict(raw);
    if (convertedTime) {
      newRow[convTimeIdx] = convertedTime;
      converted++;
    }
    outLines.push(newRow.join(sep));
  }

  const result = outLines.join('\n');
  if (outputPath) {
    fs.writeFileSync(path.resolve(outputPath), result, 'utf-8');
    console.log(`✅ ${converted} conversões processadas. Salvo em: ${outputPath}`);
  } else {
    console.log(result);
  }
}

main();
